// @sudoku/stores/history.js
import { derived, writable } from 'svelte/store';
import { createGame, createSudoku } from '../../../domain/index.js';
import { hints } from './hints';
import { grid, userGrid } from './grid';

let currentGame = createGame({ sudoku: createSudoku(initialGrid()) });
let lastHint = null;

function initialGrid() {
	return Array.from({ length: 9 }, () => Array(9).fill(0));
}

function syncUserGrid() {
	userGrid.replace(currentGame.getGrid());
}

function operationResult(ok, data = {}) {
	return { ok, ...data };
}

function toMessage(error) {
	return error instanceof Error ? error.message : String(error);
}

function conflictKeys(conflicts) {
	return conflicts.map((cell) => `${cell.col},${cell.row}`);
}

const version = writable(0);

function clearHint() {
	lastHint = null;
}

function gridSignature() {
	return currentGame.getGrid().map((row) => row.join('')).join('|');
}

function getHintIdentity(hint) {
	return `${gridSignature()}#${hint.type}#${hint.row},${hint.col}#${(hint.candidates || []).join(',')}`;
}

function nextHintLevel(identity, maxLevel = 3) {
	if (!lastHint || lastHint.key !== identity) {
		return 1;
	}

	return Math.min((lastHint.level || 1) + 1, maxLevel);
}

function consumeHint(hint, identity, repeated) {
	if (!repeated) {
		hints.useHint();
	}

	lastHint = {
		...hint,
		key: identity,
		level: hint.level || 1,
		maxLevel: hint.maxLevel || 3,
	};
	version.update((value) => value + 1);
	return operationResult(true, { hint: lastHint, repeated });
}

function withHintError(error) {
	return operationResult(false, { error: toMessage(error) });
}

function notifyChange() {
	syncUserGrid();
	version.update((value) => value + 1);
}

grid.subscribe(($grid) => {
	currentGame = createGame({ sudoku: createSudoku($grid) });
	lastHint = null;
	notifyChange();
});

export function guess(move) {
	try {
		clearHint();
		currentGame.guess(move);
		notifyChange();
		return operationResult(true, { conflicts: currentGame.getConflicts() });
	} catch (error) {
		version.update((value) => value + 1);
		return operationResult(false, { error: toMessage(error) });
	}
}

export function undo() {
	clearHint();
	currentGame.undo();
	notifyChange();
}

export function redo() {
	clearHint();
	currentGame.redo();
	notifyChange();
}

export function getCandidateHint(position) {
	try {
		const identityHint = currentGame.getCandidateHint(position.row, position.col, 2);
		const identity = getHintIdentity(identityHint);
		const repeated = Boolean(lastHint && lastHint.key === identity);
		const level = nextHintLevel(identity, identityHint.maxLevel);
		const hint = currentGame.getCandidateHint(position.row, position.col, level);
		return consumeHint(hint, identity, repeated);
	} catch (error) {
		return withHintError(error);
	}
}

export function getNextStepHint() {
	const identityHint = currentGame.getNextStepHint(2);

	if (!identityHint) {
		clearHint();
		version.update((value) => value + 1);
		return operationResult(false, { reason: 'no-single-candidate' });
	}

	const identity = getHintIdentity(identityHint);
	const repeated = Boolean(lastHint && lastHint.key === identity);
	const level = nextHintLevel(identity, identityHint.maxLevel);
	const hint = currentGame.getNextStepHint(level);
	return consumeHint(hint, identity, repeated);
}

export function applyHint(position) {
	const currentGrid = currentGame.getGrid();

	if (currentGrid[position.row][position.col] !== 0) {
		return operationResult(false, { reason: 'cell-filled' });
	}

	try {
		const hint = currentGame.getCandidateHint(position.row, position.col);

		if (hint.candidates.length !== 1) {
			return operationResult(false, { reason: 'not-single-candidate', hint });
		}

		currentGame.guess({
			row: position.row,
			col: position.col,
			value: hint.candidates[0],
		});
		lastHint = { ...hint, type: 'applied', value: hint.candidates[0] };
		notifyChange();
		return operationResult(true, { hint: lastHint, conflicts: currentGame.getConflicts() });
	} catch (error) {
		return withHintError(error);
	}
}

export function enterExplore() {
	currentGame.enterExplore();
	notifyChange();
}

export function commitExplore() {
	const committed = currentGame.commitExplore();
	notifyChange();
	return committed;
}

export function abandonExplore() {
	const abandoned = currentGame.abandonExplore();
	notifyChange();
	return abandoned;
}

export function canUndo() {
	return currentGame.canUndo();
}

export function canRedo() {
	return currentGame.canRedo();
}

export const historyState = derived(version, () => ({
	canUndo: currentGame.canUndo(),
	canRedo: currentGame.canRedo(),
}));

export const gameViewState = derived(version, () => currentGame.getViewState());

export const givenCells = derived(gameViewState, ($gameViewState) => $gameViewState.givens);

export const invalidCells = derived(gameViewState, ($gameViewState) => conflictKeys($gameViewState.conflicts));

export const exploreCells = derived(gameViewState, ($gameViewState) => conflictKeys($gameViewState.exploreCells || []));

export const gameWon = derived(gameViewState, ($gameViewState) => $gameViewState.isSolved);

export const exploreState = derived(version, () => currentGame.getExploreState());

export const hintState = derived(version, () => lastHint);
